teps/0009-trigger-crd.md

```diff
@@ -0,0 +1,269 @@
+and building pipeline.
+In this scenario, having EventListener in each namespaces or even multiple EventListener in single namespace for
+catering different use cases doesn't scale well. We get resource issue due to Pod Proliferation from EventListener.
+When this TEP is implemented, it will be the responsibility of Operator or admin to manage EventListener.
+Enduser from project will only care about Triggers and handle their webhook use cases using that.
+
+
+### Performance
+Path based EventListeners will improve the performance of EventListener by directly targeting
+the trigger definition we want to target instead of processing each trigger within a EventListener
+like we do today.
+
+At the same time, one poorly behaved customer/namespace could dominate an EventListener and affect
+other namespaces, which wouldn't be the case if each namespace had their own eventlistener.
+Operator/Admin can handle this issue by horizontally scaling EventListener or partitioning EventListener.
+
+## Design Details
+
+In first phases, this involve introduction of TriggerCRD and use of Trigger as ref inside the EventListener.
+
+Then in next phase, we will have  modification of EventListenerCRD to introduce selectors which searches
+for these Trigger resource inside of just using ref.
+EventListener will have two selectors:
+* `namespaceSelector` which specify which namespaces EventListener can search for
+Triggers. Either it can do matches based on names or labels. We will first implement this selector
+using matchName.
+* `triggerSelector` which specify which triggers can be served by EventListener. All
+triggers matching a label are served by that EventListener. This will be introduced later on.
+```
+apiVersion: v1alpha1
+kind: EventListener
+metadata:
+  name: my-el
+spec:
+  serviceAccountName: "blah"
+  namespaceSelector:
+     matchName: [ns-1, ns-2]
+     matchLabels:
+       - triggers: tekton-operator
+  triggerSelector:
+    matchLabels:
+      - eventlistener: tekton-operator
+```
+
+Trigger CR will look like the below example:
+```
+apiVersion: v1alpha1
+kind: Trigger
+metadata:
+  name: my-repo-trigger
+  labels:
+    eventlistener: tekton-operator
+spec:
+  serviceAccountName: "blah"
+  interceptors:
+    cel:
+      filter: "$(header.eventType == "push")"
+  bindings:
+      - ref: pipeline-binding
+  template:
+      name: pipeline-template
+ status:
+   address:
+      url: "el-my-svc.cluster.local" # could also be an IP address?
+   conditions:
+      type: Ready
+      status: True
+      message: "Bound to EventListener my-el"
+```
+
+Next, we will introduce Path base EventListener. Here instead of iterating through all
+the triggers, EventListener will directly process the Trigger refer by the webhook URL. If we webhook
+URL ```/bar?label=app-foo```,  EventListener will only process all the Triggers matching labels
+```app-foo```.
```

Comment:
nit: I think this messes up the GitHub render, there should be only one backtic instead of three.